src/Extension/Traits/LocalisedCopyTrait.php

```diff
@@ -0,0 +1,132 @@
+ * - localised copy of a page needs to be executed outside of usual flow
+ *
+ * @see CopyLocaleAction
+ */
+trait LocalisedCopyTrait
+{
+    /**
+     * Flag indicating if the object is going through copy to locale action
+     * this is needed to determine if we need to perform localised copy
+     * this is an in-memory only field
+     *
+     * @var bool
+     */
+    private $localisedCopyActive = false;
+
+    public function getLocalisedCopyActive(): bool
+    {
+        return $this->localisedCopyActive;
+    }
+
+    public function setLocalisedCopyActive(bool $active): self
+    {
+        $this->localisedCopyActive = $active;
+
+        return $this;
+    }
+
+    /**
+     * Provides a safe way to temporarily alter the global flag
+     * useful for unit tests
+     *
+     * @param callable $callback
+     * @return mixed
+     */
+    public function withLocalisedCopyState(callable $callback)
+    {
+        $active = $this->localisedCopyActive;
+
+        try {
+            return $callback();
+        } finally {
+            $this->setLocalisedCopyActive($active);
+        }
+    }
+
+    /**
+     * Extension point in @see CopyLocaleAction::handleAction()
+     *
+     * @param string $fromLocale
+     * @param string $toLocale
+     */
+    public function onBeforeCopyLocale(string $fromLocale, string $toLocale): void
+    {
+        $this->setLocalisedCopyActive(true);
+    }
+
+    /**
+     * Extension point in @see CopyLocaleAction::handleAction()
+     *
+     * @param string $fromLocale
+     * @param string $toLocale
+     */
+    public function onAfterCopyLocale(string $fromLocale, string $toLocale): void
+    {
+        $this->setLocalisedCopyActive(false);
+    }
+
+    /**
+     * This is executed in @see FluentExtension::onBeforeWrite()
+     */
+    public function processLocalisedCopy(): void
+    {
+        if (!$this->localisedCopyNeeded()) {
+            return;
+        }
+
+        $this->executeLocalisedCopy();
+    }
+
+    /**
+     * Determine if localised copy is needed
+     *
+     * @return bool
+     */
+    protected function localisedCopyNeeded(): bool
+    {
+        $stage = Versioned::get_stage() ?: Versioned::DRAFT;
+
+        if ($stage !== Versioned::DRAFT) {
+            // only draft stage is relevant for the duplication
+            return false;
+        }
+
+        if ($this->isInDB() && !$this->existsInLocale()) {
+            // object has a base record and doesn't have a localised record and we are localising it
+            return true;
+        }
+
+        if ($this->existsInLocale() && $this->getLocalisedCopyActive()) {
+            // object has a localised record and the content is being overridden
+            // from another locale (via copy to/from)
+            // note that we can't rely on isChanged() because writeToStage() calls forceChange()
+            // which would make this condition true every time
+            return true;
```

Comment:
@mfendeksilverstripe thanks for the succinct answers to my questions. I think this is a great change, but I feel a bit unsure about the necessity of the temporary property for flagging this functionality on/off per calls.

I wonder if we could apply a convention-over-configuration approach that simplifies the logic.

I am making the assumption that a "copy to locale" is the act of loading a record in one locale, and copying that locale's values to another.

I feel that we could possibly detect this (rather than flag this) event by inspecting query parameters.

getRecordLocale() will determine the locale that the record originated from; 

We could then use the following instead:
```php
$sourceLocale = $this->getRecordLocale();
$currentLocale = FluentState::singleton()->getLocale();
if ($this->existsInLocale() && $sourceLocale && $sourceLocale->Locale === $currentLocale) {
   // ...
```

Reply:
Note: I am assuming that even if we use a manual ->write() call, you would want the localised copy to still trigger. :)

Reply:
I might also suggest keeping the `withLocalisedCopyState` pattern, but instead of enabling this flag, you could use a similar option to globally disable the default behaviour (allowing manual migration to be performed if necessary).

E.g..

```php
$obj->withLocalisedCopyDisabled($callback);
```

Reply:
If you roll this into a yml config, you won't need a trait, so I would assume the method would be a static method on FluentExtension.

Reply:
Thanks for suggestions. Honestly, I don't like the global flag approach either so if we can get something else to do the job I would be happy to replace it. I'll test out your idea :)